hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/UpgradeDowngrade.java

```diff
@@ -130,6 +130,17 @@ public boolean needsUpgrade(HoodieTableVersion toWriteVersion) {
    * @param instantTime current instant time that should not be touched.
    */
   public void run(HoodieTableVersion toVersion, String instantTime) {
```

Comment:
Could you also audit the callers of the method and upgrade/downgrade process? I'm wondering if the `UpgradeDowngrade` should only be called if upgrade or downgrade is needed; and the caller should check auto upgrade flag. From what I read from the code:
- `BaseHoodieWriteClient#tryUpgrade` checks whether the table needs to be upgraded using `hoodie.write.table.version`.  Also note that this method takes care of rollbacks already.  So there is no need to do rollback within upgrade and downgrade handlers?